domain/application/state/application.go

```diff
@@ -2547,20 +2533,121 @@ ON CONFLICT(application_uuid) DO UPDATE SET
 	return nil
 }
 
-func encodeTrustConfig(config map[string]application.ApplicationConfig) (bool, error) {
-	trust, ok := config["trust"]
-	if !ok {
-		return false, nil
+// GetCharmConfigByApplicationID returns the charm config for the specified
+// application ID.
+// If no application is found, an error satisfying
+// [applicationerrors.ApplicationNotFound] is returned.
+// If the charm for the application does not exist, an error satisfying
+// [applicationerrors.CharmNotFoundError] is returned.
+func (st *State) GetCharmConfigByApplicationID(ctx context.Context, appID coreapplication.ID) (corecharm.ID, charm.Config, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", charm.Config{}, internalerrors.Capture(err)
+	}
+
+	appIdent := applicationID{ID: appID}
+
+	appQuery := `
+SELECT &charmUUID.*
+FROM application
+WHERE uuid = $applicationID.uuid;
+`
+	appStmt, err := st.Prepare(appQuery, appIdent, charmUUID{})
+	if err != nil {
+		return "", charm.Config{}, internalerrors.Errorf("preparing query for charm config: %w", err)
 	}
 
-	switch t := trust.Value.(type) {
-	case bool:
-		return t, nil
-	case nil:
-		return false, nil
-	default:
-		return false, internalerrors.Errorf("unexpected trust config type %T", t)
+	var (
+		ident       charmUUID
+		charmConfig charm.Config
+	)
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		if err := tx.Query(ctx, appStmt, appIdent).Get(&ident); errors.Is(err, sqlair.ErrNoRows) {
+			return applicationerrors.ApplicationNotFound
+		} else if err != nil {
+			return internalerrors.Capture(err)
+		}
+
+		charmConfig, err = st.getCharmConfig(ctx, tx, charmID{UUID: ident.UUID})
+		return internalerrors.Capture(err)
+	}); err != nil {
+		return "", charm.Config{}, internalerrors.Capture(err)
+	}
+
+	charmID, err := corecharm.ParseID(ident.UUID)
+	if err != nil {
+		return "", charm.Config{}, internalerrors.Errorf("parsing charm id: %w", err)
+	}
+
+	return charmID, charmConfig, nil
+}
+
+// GetApplicationIDByName returns the application ID for the named application.
+// If no application is found, an error satisfying
+// [applicationerrors.ApplicationNotFound] is returned.
+func (st *State) GetApplicationIDByName(ctx context.Context, name string) (coreapplication.ID, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", internalerrors.Capture(err)
+	}
+
+	var id coreapplication.ID
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		id, err = st.lookupApplication(ctx, tx, name)
+		return err
+	}); err != nil {
+		return "", internalerrors.Capture(err)
+	}
+	return id, nil
+}
```

Comment:
Does the same conversation we have had about natural keys with charms apply here?

Reply:
Yes, we need to fix this, but that can be another day.